keysign/app.py

```diff
@@ -171,61 +174,54 @@ def on_activate(self, app):
 
         window.show_all()
         self.add_window(window)
-
+        reactor.run()
 
     def run(self, args=[]):
         super(KeysignApp, self).run()
 
-    def on_key_activated(self, widget, key):
-        log.info("Activated key %r", key)
-        # Ouf, we rely on the the SendApp to have reacted to
-        # the signal first, so that it sets up the keypresentwidget
-        # and so that we can access it here.  If it did, however,
-        # We might not be able to catch the mapped signal quickly
-        # enough. So we ask the widget wether it is already mapped.
-        kpw = self.send.kpw
-        kpw.connect('map', self.on_keypresent_mapped)
-        log.debug("KPW to wait for map: %r (%r)", kpw, kpw.get_mapped())
-        if kpw.get_mapped():
-            # The widget is already visible. Let's quickly call our handler
-            self.on_keypresent_mapped(kpw)
-
-        ####
-        # Saving subtitle
-        self.headerbar_subtitle = self.headerbar.get_subtitle()
-        self.headerbar.set_subtitle("Sending {}".format(key.fpr))
-        ####
-        # Making button clickable
-        self.header_button.set_sensitive(True)
-
-
-
-
+    @staticmethod
+    def on_delete_window(*args):
+        reactor.callFromThread(reactor.stop)
+        Gtk.main_quit(*args)
 
     def on_sr_stack_switch(self, stack, *args):
         log.debug("Switched Stack! %r", args)
         #self.update_header_button()
 
     def on_send_stack_switch(self, stack, *args):
         log.debug("Switched Send Stack! %r", args)
-        #self.update_header_button()
+        current = self.send.stack.get_visible_child()
+        if current == self.send.klw:
+            log.debug("Key list page now visible")
+            self.on_keylist_mapped(self.send.klw)
+        elif current == self.send.kpw:
+            log.debug("Key present page now visible")
+            self.on_keypresent_mapped(self.send.kpw)
+        elif current == self.send.rb:
+            log.debug("Result page now visible")
+            self.on_resultbox_mapped(self.send.rb)
 
```

Comment:
Now we can't call `kpw.connect()` when we are in `on_key_activated` of app.py, because when `on_key_activated` is executed the kpw object still don't exist.
This is due to the fact that before we generate a kpw object here https://github.com/gnome-keysign/gnome-keysign/pull/15/files#diff-554ea0006765f03a4424c0c27a5550bcL85
But now I moved it in the `on_code_generated` of send.py https://github.com/gnome-keysign/gnome-keysign/pull/15/files#diff-554ea0006765f03a4424c0c27a5550bcR123

And on_code_generated will be most likely executed after the `on_key_activated` of app.py

I don't know if I explained it better.